File: hecho con java/src/main/java/com/umanizales/notasya/services/AnaliticaService.java
package com.umanizales.notasya.services;

import com.umanizales.notasya.dtos.ResumenAcademicoDTO;
import com.umanizales.notasya.models.Curso;
import com.umanizales.notasya.patterns.strategy.AcademicEvaluatorContext;
import com.umanizales.notasya.patterns.strategy.HonorsGradingStrategy;
import com.umanizales.notasya.repositories.CursoRepository;
import com.umanizales.notasya.repositories.EstudianteRepository;
import com.umanizales.notasya.repositories.ProfesorRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class AnaliticaService {

    private final EstudianteRepository estudianteRepository;
    private final ProfesorRepository profesorRepository;
    private final CursoRepository cursoRepository;
    private final AcademicEvaluatorContext evaluator;

    public AnaliticaService(EstudianteRepository estRepo, ProfesorRepository profRepo, CursoRepository cursoRepo) {
        this.estudianteRepository = estRepo;
        this.profesorRepository = profRepo;
        this.cursoRepository = cursoRepo;
        this.evaluator = new AcademicEvaluatorContext(new HonorsGradingStrategy());
    }

    @Transactional(readOnly = true)
    public ResumenAcademicoDTO obtenerResumen() {
        long totalEst = estudianteRepository.count();
        long totalProf = profesorRepository.count();
        List<Curso> cursos = cursoRepository.findAll();
        long totalCursos = cursos.size();

        if (totalCursos == 0) {
            Map<String, Integer> emptyMap = new HashMap<>();
            emptyMap.put("Aprobados", 0);
            emptyMap.put("Reprobados", 0);
            return ResumenAcademicoDTO.builder()
                    .totalEstudiantes(totalEst)
                    .totalProfesores(totalProf)
                    .totalCursos(0)
                    .promedioGeneral(0.0)
                    .tasaAprobacionPorcentaje(0.0)
                    .estudiantesConHonor(0)
                    .cursosPorEstado(emptyMap)
                    .build();
        }

        double suma = cursos.stream().mapToDouble(Curso::getCalificacion).sum();
        double promedio = Math.round((suma / totalCursos) * 100.0) / 100.0;

        int aprobados = 0;
        int conHonor = 0;
        int reprobados = 0;

        for (Curso c : cursos) {
            Map<String, Object> eval = evaluator.executeEvaluation(c.getCalificacion());
            boolean isHonor = (boolean) eval.get("honors");
            if (isHonor) {
                conHonor++;
                aprobados++;
            } else if (c.getCalificacion() >= 3.0) {
                aprobados++;
            } else {
                reprobados++;
            }
        }

        double tasa = Math.round(((double) aprobados / totalCursos) * 1000.0) / 10.0;

        Map<String, Integer> estados = new HashMap<>();
        estados.put("Aprobados", aprobados);
        estados.put("Reprobados", reprobados);
        estados.put("Con Mención de Honor", conHonor);

        return ResumenAcademicoDTO.builder()
                .totalEstudiantes(totalEst)
                .totalProfesores(totalProf)
                .totalCursos(totalCursos)
                .promedioGeneral(promedio)
                .tasaAprobacionPorcentaje(tasa)
                .estudiantesConHonor(conHonor)
                .cursosPorEstado(estados)
                .build();
    }
}

File: hecho con python/app/services/analitica_service.py
from sqlalchemy.orm import Session
from app.models.estudiante import Estudiante
from app.models.profesor import Profesor
from app.models.curso import Curso
from app.patterns.strategy import HonorsGradingStrategy, AcademicEvaluatorContext
from app.dtos.analitica_dto import ResumenAcademicoDTO

class AnaliticaAcademicaService:
    """
    Servicio de inteligencia analítica académica.
    Aplica el patrón Strategy para evaluar el rendimiento global de la institución.
    """
    def __init__(self, db: Session):
        self.db = db
        self.evaluator = AcademicEvaluatorContext(HonorsGradingStrategy())

    def obtener_resumen_academico(self) -> ResumenAcademicoDTO:
        total_est = self.db.query(Estudiante).count()
        total_prof = self.db.query(Profesor).count()
        cursos = self.db.query(Curso).all()
        total_cursos = len(cursos)

        if total_cursos == 0:
            return ResumenAcademicoDTO(
                total_estudiantes=total_est,
                total_profesores=total_prof,
                total_cursos=0,
                promedio_general=0.0,
                tasa_aprobacion_porcentaje=0.0,
                estudiantes_con_honor=0,
                cursos_por_estado={"Aprobado": 0, "Reprobado": 0, "Mención de Honor": 0}
            )

        suma_notas = sum(c.calificacion for c in cursos)
        promedio = round(suma_notas / total_cursos, 2)

        aprobados = 0
        con_honor = 0
        reprobados = 0

        for c in cursos:
            res = self.evaluator.execute_evaluation(c.calificacion)
            if res["honors"]:
                con_honor += 1
                aprobados += 1
            elif res["grade"] >= 3.0:
                aprobados += 1
            else:
                reprobados += 1

        tasa_aprob = round((aprobados / total_cursos) * 100, 1)

        return ResumenAcademicoDTO(
            total_estudiantes=total_est,
            total_profesores=total_prof,
            total_cursos=total_cursos,
            promedio_general=promedio,
            tasa_aprobacion_porcentaje=tasa_aprob,
            estudiantes_con_honor=con_honor,
            cursos_por_estado={
                "Aprobados": aprobados,
                "Reprobados": reprobados,
                "Con Mención de Honor": con_honor
            }
        )
